File: src/main/GO/leetcode/lists/classic/Classic1618.go
package classic

import (
	"strings"
)

/**
 * @Author: jingxinwu
 * @Date: 2023/2/17 12:55
 */

func patternMatching(pattern string, value string) bool {
	if len(pattern) == 0 {
		return len(value) == 0
	}
	if len(value) == 0 {
		return len(pattern) == 1
	}
	count := func(ch byte) int {
		sum := 0
		for i := 0; i < len(pattern); i++ {
			if ch == pattern[i] {
				sum++
			}
		}
		return sum
	}

	isEqual := func(s1 string, offset1, offset2, size int) bool {
		for i := 0; i < size; i++ {
			if s1[offset1+i] != s1[offset2+i] {
				return false
			}
		}
		return true
	}
	matches := func(aSize int, bIndex, bSize int) bool {
		index := aSize
		for i := 1; i < len(pattern); i++ {
			var size, offset int
			if pattern[i] == pattern[0] {
				size = aSize
				offset = 0
			} else {
				size = bSize
				offset = bIndex
			}
			if !isEqual(value, index, offset, size) {
				return false
			}
			index += size
		}
		return true
	}

	var aCh, bCh byte
	aCh = pattern[0]

	if aCh == 'a' {
		bCh = 'b'
	} else {
		bCh = 'a'
	}
	size := len(value)

	//一旦选定了 a，我们也就相应的选定了 b
	countA := count(aCh)
	countB := len(pattern) - countA
	firstB := strings.Index(pattern, string(bCh))

	for aSize := 3; aSize <= size/countA; aSize++ {
		remainingLength := size - aSize*countA
		if countB == 0 || remainingLength%countB == 0 {
			bIndex := firstB * aSize
			bSize := 0
			if countB != 0 {
				bSize = remainingLength / countB
			}
			if matches(aSize, bIndex, bSize) {
				if value[0:aSize] != value[aSize:aSize+bSize] {
					return true
				}
			}
		}
	}
	return false
}

File: java/src/leetcode/lists/classic/Classic1618.java
package leetcode.lists.classic;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/8/31 14:20
 */
public class Classic1618 {

    public boolean patternMatching(String pattern, String value) {
        if (pattern.isEmpty()) {
            return value.isEmpty();
        }

        if (value.isEmpty()) {
            return pattern.length() == 1;
        }

        char mainChar = pattern.charAt(0);
        char altChar = mainChar == 'a' ? 'b' : 'a';
        int size = value.length();

        int countOfMain = countOf(pattern, mainChar);
        int countOfAlt = pattern.length() - countOfMain;
        int firstAlt = pattern.indexOf(altChar);
        int maxMainSize = size / countOfMain;

        for (int mainSize = 0; mainSize <= maxMainSize; mainSize++) {
            int remainingLength = size - mainSize * countOfMain;
            if (countOfAlt == 0 || remainingLength % countOfAlt == 0) {
                int altIndex = firstAlt * mainSize;
                int altSize = countOfAlt == 0 ? 0 : remainingLength / countOfAlt;
                if (matches(pattern, value, mainSize, altSize, altIndex)) {

//                    System.out.println(value.substring(0, mainSize));
//                    System.out.println(value.substring(mainSize, mainSize + altSize));

                    if (!value.substring(0, mainSize).equals(value.substring(mainSize, mainSize + altSize))) {
                        return true;

                    }
                }
            }
        }
        return false;
    }

    public static boolean matches(String pattern, String value, int mainSize, int altSize, int firstAlt) {
        int stringIndex = mainSize;
        for (int i = 1; i < pattern.length(); i++) {
            int size = pattern.charAt(i) == pattern.charAt(0) ? mainSize : altSize;
            int offset = pattern.charAt(i) == pattern.charAt(0) ? 0 : firstAlt;
            if (!isEqual(value, offset, stringIndex, size)) {
                return false;
            }
            stringIndex += size;
        }
        return true;
    }

    public static boolean isEqual(String s1, int offset1, int offset2, int size) {
        for (int i = 0; i < size; i++) {
            if (s1.charAt(offset1 + i) != s1.charAt(offset2 + i)) {
                return false;
            }
        }
        return true;
    }

    public static int countOf(String pattern, char c) {
        int count = 0;
        for (int i = 0; i < pattern.length(); i++) {
            if (pattern.charAt(i) == c) {
                count++;
            }
        }
        return count;
    }
}
